crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,957 @@
+    /// Generate a new SlidingSyncBuilder with the same inner settings and views
+    /// but without the current state
+    pub fn new_builder_copy(&self) -> SlidingSyncBuilder {
+        let mut builder = SlidingSyncBuilder::default()
+            .client(self.client.clone())
+            .views(Arc::new(futures_signals::signal_vec::MutableVec::new_with_values(
+                self.views
+                    .lock_ref()
+                    .to_vec()
+                    .iter()
+                    .map(|v| {
+                        v.new_builder().build().expect("builder worked before, builder works now")
+                    })
+                    .collect(),
+            )))
+            .subscriptions(Arc::new(futures_signals::signal_map::MutableBTreeMap::with_values(
+                self.subscriptions.lock_ref().to_owned(),
+            )))
+            .to_owned();
+
+        if let Some(ref h) = self.homeserver {
+            builder.homeserver(h.clone());
+        }
+        builder
+    }
+
+    /// Subscribe to a given room.
+    ///
+    /// Note: this does not cancel any pending request, so make sure to only
+    /// poll the stream after you've altered this. If you do that during, it
+    /// might take one round trip to take effect.
+    pub fn subscribe(&self, room_id: OwnedRoomId, settings: Option<v4::RoomSubscription>) {
+        self.subscriptions.lock_mut().insert_cloned(room_id, settings.unwrap_or_default());
+    }
+
+    /// Unsubscribe from a given room.
+    ///
+    /// Note: this does not cancel any pending request, so make sure to only
+    /// poll the stream after you've altered this. If you do that during, it
+    /// might take one round trip to take effect.
+    pub fn unsubscribe(&self, room_id: OwnedRoomId) {
+        if self.subscriptions.lock_mut().remove(&room_id).is_some() {
+            self.unsubscribe.lock_mut().push_cloned(room_id);
+        }
+    }
+
+    /// Lookup a specific room
+    pub fn get_room(&self, room_id: OwnedRoomId) -> Option<SlidingSyncRoom> {
+        self.rooms.lock_ref().get(&room_id).cloned()
+    }
+
+    /// Lookup a set of rooms
+    pub fn get_rooms<I: Iterator<Item = OwnedRoomId>>(
+        &self,
+        room_ids: I,
+    ) -> Vec<Option<SlidingSyncRoom>> {
+        let rooms = self.rooms.lock_ref();
+        room_ids.map(|room_id| rooms.get(&room_id).cloned()).collect()
+    }
+
+    async fn handle_response(
+        &self,
+        resp: v4::Response,
+        views: &[SlidingSyncView],
+    ) -> anyhow::Result<UpdateSummary> {
+        self.client.process_sliding_sync(resp.clone()).await?;
+        tracing::info!("main client processed.");
+        self.pos.replace(Some(resp.pos));
+        let mut updated_views = Vec::new();
+        if resp.lists.len() != views.len() {
+            bail!("Received response for {} lists, yet we have {}", resp.lists.len(), views.len());
+        }
+
+        for (view, updates) in views.iter().zip(resp.lists.iter()) {
```

Comment:
```suggestion
        for (view, updates) in iter::zip(&views, &resp.lists) {
```

plus `use std::iter;` would be a little shorter / easier to read.